### PR TITLE
feat(cell_based_table): add sentinel column and cell_based muti_get

### PR DESCRIPTION
**Pull request by wcy-fdu**
## What's changed and what's your intention?

This PR first adds sentinel column when serialize/deserialize, using None value for sentinel cell. While state_store is getting  sentinel column, there are three cases:
- `return None`: no exist row
- `return Some(value)`: exist row

Then cell_based `muti_get` is implemented by the design above.

BTW, some unit tests have been modified due to corresponding changes.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

### CODE REVIEW

File: src/common/src/util/ordered/serde.rs [hunk 1]
```diff
@@ -154,32 +153,27 @@ pub fn serialize_pk_and_row(
...
-        // gets nothing.
-        let key = [
-            pk_buf,
-            serialize_column_id(&NULL_ROW_SPECIAL_CELL_ID)?.as_slice(),
-        ]
-        .concat();
+
+    if row.is_none() {
+        let key = [pk_buf, serialize_column_id(&SENTINEL_CELL_ID)?.as_slice()].concat();
+        result.push((key, None));
```
Comment (skyzh) [inline src/common/src/util/ordered/serde.rs:170]: The signature of `serialize_pk_and_row` should be changed in the future. As all deletions are now bound with the original value, there's no possibility of passing a `None` row into this function.

File: src/storage/src/cell_based_row_deserializer.rs [hunk 2]
```diff
@@ -62,18 +61,16 @@ impl CellBasedRowDeserializer {
         let pk_vec_len = pk_with_cell_id.len();
         let cur_pk_bytes = &pk_with_cell_id[0..pk_vec_len - 4];
         let mut result = None;
-        if let Some(prev_pk_bytes) = &self.pk_bytes && prev_pk_bytes != cur_pk_bytes {
+        let cell_id_bytes = &pk_with_cell_id[pk_vec_len - 4..];
```
Comment (skyzh) [inline src/storage/src/cell_based_row_deserializer.rs:64]: Better to check length here. If key length < 4, return error.

Also, consider using `.split_at` function to create two slices.

File: src/storage/src/table/cell_based_table.rs [hunk 3]
```diff
@@ -125,9 +125,47 @@ impl<S: StateStore> CellBasedTable<S> {
     }
 
     // cell-based interface
-    pub async fn get_row(&self, _pk: &Row, _epoch: u64) -> StorageResult<Option<Row>> {
-        // get row by state_store multi get
-        todo!()
+    pub async fn get_row(&self, pk: &Row, epoch: u64) -> StorageResult<Option<Row>> {
+        let pk_serializer = self.pk_serializer.as_ref().expect("pk_serializer is None");
+        let column_ids = generate_column_id(&self.column_descs);
```
Comment (skyzh) [inline src/storage/src/table/cell_based_table.rs:130]: Why not pre-generate `column_ids` when calling CellBasedTable::new?

File: src/storage/src/table/cell_based_table.rs [hunk 4]
```diff
@@ -125,9 +125,47 @@ impl<S: StateStore> CellBasedTable<S> {
...
+        // let mut row = Vec::with_capacity(column_ids.len());
+        let sentinel_key = [
+            &serialize_pk(pk, pk_serializer).map_err(err)?[..],
```
Comment (skyzh) [inline src/storage/src/table/cell_based_table.rs:133]: pk only needs to be serialized once.

File: src/storage/src/table/cell_based_table.rs [hunk 5]
```diff
@@ -125,9 +125,47 @@ impl<S: StateStore> CellBasedTable<S> {
...
+                &serialize_column_id(&column_id).map_err(err)?,
+            ]
+            .concat();
+            let state_store_get_res = self.keyspace.get(&key, epoch).await?;
+            if let Some(state_store_get_res) = state_store_get_res {
+                get_res.push((key, state_store_get_res));
+            }
+        }
+        let mut cell_based_row_deserializer =
+            CellBasedRowDeserializer::new(self.column_descs.clone());
```
Comment (skyzh) [inline src/storage/src/table/cell_based_table.rs:158]: May optimize `CellBasedRowDeserializer` to take reference in the future, or refactor it to take `Arc<[ColumnDesc]>`.

File: src/stream/src/executor_v2/mview/materialize.rs [hunk 6]
```diff
@@ -235,7 +235,7 @@ mod tests {
                         .await
                         .unwrap()
                         .len(),
-                    6
+                    9
```
[the remark this answers is not in the document]
Comment (skyzh) [reply src/stream/src/executor_v2/mview/materialize.rs:238]: also update comments above

File: src/stream/src/executor/managed_state/top_n/top_n_state.rs [hunk 7]
```diff
@@ -298,7 +298,7 @@ impl<S: StateStore, const TOP_N_TYPE: usize> ManagedTopNState<S, TOP_N_TYPE> {
         let pk_row_bytes = self
             .keyspace
             .scan_strip_prefix(
-                number_rows.map(|top_n_count| top_n_count * self.data_types.len()),
+                number_rows.map(|top_n_count| top_n_count * (self.data_types.len() + 1)),
```
[the remark this answers is not in the document]
Comment (lmatz) [reply src/stream/src/executor/managed_state/top_n/top_n_state.rs:301]: Yes, I talked to @wcy-fdu about this part requiring refactoring in the future.

File: src/stream/src/executor_v2/mview/table_state_tests.rs [hunk 8]
```diff
@@ -499,3 +499,78 @@ async fn test_get_row_by_scan() {
         .unwrap();
     assert_eq!(get_no_exist_res, None);
 }
+
+#[tokio::test]
+async fn test_get_row_by_muti_get() {
+    let state_store = MemoryStateStore::new();
+    let column_ids = vec![ColumnId::from(0), ColumnId::from(1), ColumnId::from(2)];
```
[the remark this answers is not in the document]
Comment (wcy-fdu) [reply src/stream/src/executor_v2/mview/table_state_tests.rs:506]: Sounds good, I will add these cases and remove `get_for_test` in cell_based_table.

**Review by skyzh — APPROVED** [review]
LGTM. Better also have a test case like write using `1, 2, 3` column id, and read using `3, 2`.

**Comment by wcy-fdu** [comment]
> LGTM. Better also have a test case like write using `1, 2, 3` column id, and read using `3, 2`.

Good idea.

**Pull request merged**